### PR TITLE
Story/parameter tabs/2737

### PR DESCRIPTION
**Pull request by conbrad**
- `ColumnDefBuilder.tsx` builds the column definitions for the data grid
- `GridComponentRenderer.tsx` renders the components of the data grid (cells, etc.)
- `TabbedDataGrid` handles the transition between weather parameter tab data grid or the forecast summary data grid:

   - image
    - a copy of the redux state is held as local state in `TabbedDataGrid` to pass down to the 2 data grids and maintain state between them
- `Morecast2Row` holds all the combinations of forecast/actual/prediction values available for display/update from anywhere within the data grids.

# Test Links:
[Landing Page](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-2801.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: web/src/features/moreCast2/saveForecast.test.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+          buildCompleteForecast('1', mockForDate, 1, 'one'),
+          buildInvalidForecast('2', mockForDate, 2, 'two')
+        ])
+      ).toBe(false)
+    })
+
+    it('should return false if any forecasts have missing forecast fields set other than windDirectionForecast', () => {
+      expect(
+        isForecastValid([
+          buildCompleteForecast('1', mockForDate, 1, 'one'),
```
Comment (andrea-williams) [inline web/src/features/moreCast2/saveForecast.test.ts:160]: shouldn't this be calling `buildForecastMissingWindDirection()` instead?

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+      })
+      // The most recent value from the weather station for the weather parameter of interest
+      // (eg. tempActual) which will be applied as the forecast value
+      const mostRecentValue = mostRecentRow[actualField]
+      // Finally, get an array of rows that don't have an actual for the weather parameter of interest
+      // and iterate through them to apply the most recent actual as the new forecast value
+      const rowsWithoutActuals: MoreCast2Row[] = values.filter(value => isNaN(value[actualField] as number))
+      rowsWithoutActuals.forEach(row => {
+        const predictionItem = row[forecastField] as PredictionItem
+        predictionItem.choice = ModelChoice.YESTERDAY
```
Comment (andrea-williams) [inline web/src/features/moreCast2/components/TabbedDataGrid.tsx:231]: might want to consider changing the name of this to `ModelChoice.PERSISTENCE` - if we're using the term "persistence" on the FE & with users, I suspect after a while we'll forget what "Yesterday" even means when we return to this code later

File: web/src/features/moreCast2/slices/dataSlice.ts [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+  const dateInterval = createDateInterval(fromDate, toDate)
+  const groupedByStationCode = createStationCodeToWeatherIndeterminateGroups(items, stationMap)
+
+  const allPredictions = [...items]
+  for (const [key, values] of Object.entries(groupedByStationCode)) {
+    const stationCode = parseInt(key)
+    const stationName = stationMap.get(stationCode) || ''
+    const groupedByUtcTimestamp = createUtcTimeStampToWeatherIndeterminateGroups(values, dateInterval)
+
+    for (const [key2, values2] of Object.entries(groupedByUtcTimestamp)) {
```
Comment (andrea-williams) [inline web/src/features/moreCast2/slices/dataSlice.ts:306]: could we pick better variable names than key/key2, values/values2 please?

**Review by andrea-williams — COMMENTED** [review]
I think I'm done reviewing the code. Haven't looked at it functioning in dev yet...

File: web/tsconfig.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "target": "es5",
+    "target": "es2015",
```
[the remark this answers is not in the document]
Comment (conbrad) [reply web/tsconfig.json:3]: Yes, without it we can't use the `keys()` method on a map (we do this in a few places in `dataSlice.ts`), we get the error:

```
TS2802: Type 'IterableIterator<number>' can only be iterated through when using the '--downlevelIteration' flag or with a '--target' of 'es2015' or higher.
```
And `es2015` has been supported across all major browsers a million versions ago.

File: web/src/features/moreCast2/components/applyToColumnMenu.test.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        contextMenu={{ mouseX: 1, mouseY: 1 }}
+        updateColumnWithModel={mockUpdateColumnWithModel}
+        handleClose={mockHandleClose}
+      />
+    )
+    const menu = getByTestId('apply-to-column-menu')
+    expect(menu).toBeInTheDocument()
+  })
+
+  it('should render the menu when context from a click exists', async () => {
```
[the remark this answers is not in the document]
Comment (conbrad) [reply web/src/features/moreCast2/components/applyToColumnMenu.test.tsx]: Thanks, removed

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx (hunk 2 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/TabbedDataGrid.tsx]: Done!

**Review by dgboss — APPROVED** [review]

**Review by andrea-williams — APPROVED** [review]
Observations (I wouldn't go so far as to say that these are "broken"):

- when you go crazy and select all the weather tabs, the station name column gets smushed down to only the first 2 letters, and the date column gets smushed down to 1 letter. In the future we'll probably want to keep the widths of these 2 columns pretty static and probably freeze them so that when you're scrolling across the width of the page you know what station & date you're looking at in each row
- it's possible to deselect all the tabs, then the table just has rows for station & date. It's potentially a little confusing for new users
- the top-level header for the Station & Date cols is labelled "ID" but I don't think it serves any purpose - we could probably remove that label?

I don't recommend that we implement any of these changes in this PR - follow-up tickets instead.

Let's ship this!

**Pull request merged**